Rewrite {recip,rsqrt,sqrt} LLKs.

jasondavies opened the pull request:
See:

- tenstorrent/tt-metal#21622

The original implementations used poor initial guesses for Newton-Raphson, resulting in poor results even when using 3 iterations.

These new implementations use state-of-the-art constants for the initial guesses, along with algorithmic improvements.

The new implementations have ~1 ulps error for either approximate mode (10 bits) or non-approximate mode (~23 bits).

The sqrt LLK now also has support for optional rsqrt (reciprocal sqrt), which uses the same number of multiplies as sqrt.

### Ticket
<!-- Link to Github Issue -->

### Problem description
<!-- Provide context for the problem. -->

### What's changed
<!-- Describe the approach used to solve the problem.
Summarize the changes made and its impact. -->

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist
<!-- These are required steps and need to be run from tt-metal repository's Actions. Use links below and replace them with your run -->
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)

jasondavies commented on tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_recip.h:
```diff
@@ -11,61 +12,37 @@ namespace ckernel
+        sfpi::vFloat out = _sfpu_reciprocal_<APPROXIMATION_MODE>(in);
 
         if constexpr (is_fp32_dest_acc_en || APPROXIMATION_MODE)
         {
```
Can anyone explain why this is here?

jasondavies replied:
By "this", I mean the line below (that I can't seem to add a comment to), that says:

```cpp
sfpi::dst_reg[0] = sfpi::reinterpret<sfpi::vFloat>(float_to_fp16b(out, 0));
```

This happens when is_fp32_dest_acc_en=false and APPROXIMATION_MODE=false.

jasondavies replied:
See [here](https://github.com/tenstorrent/tt-llk/commit/9f00b53fe5583dc8946c4e66c89b127e37d06584#r157770862) for where it got added by @rdjogoTT.

rdjogoTT replied:
fp16b formats in Dest were having accuracy issues because the recip is computed in fp32 in the LREGs, but when we store datums back to Dest from LREGs it truncates the lower bits. This was added to do rounding before storing values back to Dest, and helped with the accuracy problem. We do have an internal issue open to potentially look at adding this to other places as well, it's just low priority right now.

jasondavies replied:
In an ideal world it would probably be good to support rounding to both bf16 and fp16, as this assumes bf16 right now.

fvranicTT commented:
@jasondavies Can you try adding the tests following the same approach as in https://github.com/tenstorrent/tt-llk/pull/101?
See https://github.com/tenstorrent/tt-llk/blob/main/tests/README.md on how to run the tests. One note: at the moment automatic detection of cards is broken, I'm looking into it, but it should work if you `export CHIP_ARCH=wormhole`.
These tests are checking pcc, not ideal thing, but should be enough for now. We plan to start checking the ULPs, but we're not there yet.

jasondavies commented:
I'm confused why Sin/Cos tests are failing as I haven't touched them:

```
>           assert torch.isclose(
                golden_tensor[i], res_tensor[i], rtol=rtol, atol=atol
            ), f"Failed at index {i} with values {golden[i]} and {res_from_L1[i]}"
E           AssertionError: Failed at index 0 with values 0.937653277514828 and 131008.0
E           assert tensor(False)
E            +  where tensor(False) = <built-in method isclose of type object at 0x7f2937f86bc0>(tensor(0.94, dtype=torch.bfloat16), tensor(131072., dtype=torch.bfloat16), rtol=0.1, atol=0.05)
E            +    where <built-in method isclose of type object at 0x7f2937f86bc0> = torch.isclose

test_eltwise_unary_sfpu.py:169: AssertionError
```

fvranicTT commented:
> I'm confused why Sin/Cos tests are failing as I haven't touched them:
> 
> ```
> >           assert torch.isclose(
>                 golden_tensor[i], res_tensor[i], rtol=rtol, atol=atol
>             ), f"Failed at index {i} with values {golden[i]} and {res_from_L1[i]}"
> E           AssertionError: Failed at index 0 with values 0.937653277514828 and 131008.0
> E           assert tensor(False)
> E            +  where tensor(False) = <built-in method isclose of type object at 0x7f2937f86bc0>(tensor(0.94, dtype=torch.bfloat16), tensor(131072., dtype=torch.bfloat16), rtol=0.1, atol=0.05)
> E            +    where <built-in method isclose of type object at 0x7f2937f86bc0> = torch.isclose
> 
> test_eltwise_unary_sfpu.py:169: AssertionError
> ```
> 
> Edit: looks like they break on main too, so nothing to do with me!

That’s odd. They were passing yesterday.

fvranicTT commented:
Let's disable it for now. I can take care of that.

jasondavies commented:
I've only disabled the Cos tests, which were failing for some reason.  Still quite confused by what's causing that.

amahmudTT commented on tests/sources/eltwise_unary_sfpu_test.cpp:
```diff
@@ -63,12 +63,16 @@ void call_sfpu_operation(SfpuType operation)
             ckernel::sfpu::_calculate_reciprocal_<APPROX_MODE, 10, is_fp32_dest_acc_en>(10);
             break;
+        case SfpuType::rsqrt:
+            ckernel::sfpu::_init_sqrt_<APPROX_MODE>();
```
Instead of using the same functions should we have a wrapper for rsqrt that should call the sqrt functions with the proper template arguments ? For example the initial function for both of them are the same, might be helpful if they are different. 

In other words we could have _calculate_sqrt_internal or something that has the extra boolean template argument and is called by _calculate_sqrt_ or _calculate_rsqrt_.

jasondavies replied:
Seems reasonable.  Done!

amahmudTT commented on tt_llk_blackhole/common/inc/sfpu/ckernel_sfpu_recip.h:
```diff
@@ -81,8 +59,9 @@ inline void _calculate_reciprocal_(const int iterations)
 {
-    sfpi::vConstFloatPrgm0 = 1.442695f; // ln2_recip
-    sfpi::vConstFloatPrgm1 = 2.0f;
+    sfpi::vConstIntPrgm0   = 0x7eb504f3;
```
Do these numbers have special meaning ? For example in the deleted code the constant for 1/ln2 had a comment beside it. If these numbers have such meanings then corresponding comments will be helpful while going over the relevant paper (if anyone needs to go through it sometime).

amahmudTT commented on tt_llk_blackhole/common/inc/sfpu/ckernel_sfpu_sqrt.h:
```diff
@@ -12,48 +12,60 @@ namespace ckernel
-            // approximation of square root
-            val_s >>= 1;
-            sfpi::dst_reg[0] = sfpi::reinterpret<sfpi::vFloat>(val_s);
+            y = y * (sfpi::vConstFloatPrgm1 + negative_y * c);
```
Curious , what is the reason of having a dedicated negative_y variable ? To make sure MAD instructions are used or SFPI does not support subtraction ?

jasondavies replied:
I must admit, I don't particularly trust SFPI at the moment to do what I would expect w.r.t optimisation.  In the above example, _subtraction_, i.e. `sfpi::vConstFloatPrgm1 - y * c`, actually generates a whole extra SFPMUL (`-1.0 * y`).

There is also somewhat of a case for readability with optimisation in mind: the preceding operation (`c = x * y`) takes 2 cycles, which means a NOP is required unless the next instruction is non-dependent, which luckily for us, `negative_y = -y` is (which emits a 1-cycle SFPMOV with negation flag).  In other words, this is what SFPI _should_ (probably) emit here, and making it explicit is a little clearer for anyone reading it with an optimisation mindset.

Note also that the code was written primarily with Wormhole in mind, where MAD does not support negation of the multiplication or addend.  I believe this is supported in Blackhole, so there may be some room to improve marginally here, but I don't have access to confirm.

amahmudTT commented on tt_llk_blackhole/common/inc/sfpu/ckernel_sfpu_sqrt.h:
```diff
@@ -63,11 +75,14 @@ inline void _init_sqrt_()
     if (APPROXIMATION_MODE)
     {
-        sfpi::vConstFloatPrgm0 = sfpi::s2vFloat16b(127 << 7);
+        sfpi::vConstIntPrgm0   = 0x5f0b3892;
```
Same comment regarding these values having special meanings and addition of corresponding comments.

amahmudTT commented on tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_binary.h:
```diff
@@ -167,7 +167,7 @@ inline void _sfpu_binary_init_()
     if constexpr (BINOP == BinaryOp::DIV || BINOP == BinaryOp::POW)
     {
-        _init_reciprocal_<APPROXIMATION_MODE>();
+        _init_reciprocal_<false>();
```
Are we intentionally ignoring the APPROXIMATE_MODE ?

jasondavies replied:
Yes: this initialisation is specifically for the use of `_sfpu_reciprocal_` in either DIV/POW.  In both cases the default param (`APPROXIMATE=false`) is used.

amahmudTT commented on tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_exp.h:
```diff
@@ -360,9 +360,7 @@ inline void _init_exponential_()
-        sfpi::vConstFloatPrgm0 = 1.442695f; // ln2_recip
-        sfpi::vConstFloatPrgm1 = 2.0f;
-        sfpi::vConstFloatPrgm2 = 0.863281f;
+        _init_reciprocal_<false>();
```
What are the heuristics/reasonings for choosing approximate mode vs non-approximate mode ? You could make it clear in a comment somewhere.

jasondavies replied:
This initialisation sets constants for the use of _sfpu_reciprocal_<false>.  Happy to add a comment explaining this.

amahmudTT commented on tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_recip.h:
```diff
@@ -83,8 +60,9 @@ inline void _calculate_reciprocal_(const int iterations)
 {
-    sfpi::vConstFloatPrgm0 = 1.442695f; // ln2_recip
-    sfpi::vConstFloatPrgm1 = 2.0f;
+    sfpi::vConstIntPrgm0   = 0x7eb504f3;
```
Same comment as before.

amahmudTT commented on tt_llk_wormhole_b0/common/inc/sfpu/ckernel_sfpu_sqrt.h:
```diff
@@ -12,48 +13,60 @@ namespace ckernel
+        }
+        else
+        {
+            half_y = -half_y;
```
Why flip sign ? If it is for using fused multiply and add, did you check if it does not work when using subtraction ? By dumping the assembly ? If it works but the compiler is not smart enough then it might just be negating the number and wasting cycles.

jasondavies commented:
> an explanation as to why we use approximate mode and why not in certain cases (where there is no definite rule) would be helpful

I'm personally not a big fan of the way this vague "approximate mode" flag (and even more confusingly, "fast approx" for exp) is used everywhere but only actually used in some places.

Specifically here, I'm calling the initialisation function for whatever function is being used in the main body.  Which seems to be `_sfpu_reciprocal_<APPROXIMATE = false>` (again, lots of confusion thanks to default template parameters etc.)

I've added comments that hopefully make this clearer.

amahmudTT commented:
LGTM. Thanks for addressing the comments. Will approve once tests pass, I am assuming all post commit tests and all post commit blackhole tests will be run.

jasondavies commented:
> I am assuming all post commit tests and all post commit blackhole tests will be run.

Thanks.  Do you mean the tt-metal integration tests?  These won't pass without tenstorrent/tt-metal#21623 - this is due to hardcoded constants being present in tt-metal.

jasondavies commented:
Integration tests failed as expected:

1. We have duplicate trig code that causes all integration tests with tt-metal main to fail right now (I've had to add a workaround that removes the duplicate code in some of my tt-metal PRs).
2. _This_ PR specifically requires the changes in tenstorrent/tt-metal#21623 (removes hardcoded constants in tt-metal).

fvranicTT commented:
@jasondavies Conflicts between metal and llk have been solved. Could you please rebase? Thank you for your patience.

fvranicTT commented:
@jasondavies Got it. But we also have another breaking change that just landed in the llk:
https://github.com/tenstorrent/tt-llk/pull/282. This one needs to go in a lockstep with https://github.com/tenstorrent/tt-metal/pull/22860
Let’s wait for it and then merge your changes in another lockstep.

jasondavies commented:
LLK tests are green, this should be ready to merge.

Integration tests on this PR will fail as this needs https://github.com/tenstorrent/tt-metal/pull/21623 to pass - maybe integration tests labels should be removed.

We can run tests on the tt-metal PR once this is merged to main.

ldjurovicTT commented:
@jasondavies As far as tests LLK are considered I approved. Let's just see what @fvranicTT says about integration with metal

fvranicTT commented:
> @jasondavies As far as tests LLK are considered I approved. Let's just see what @fvranicTT says about integration with metal

That's clear. Integration tests will fail without the metal change. I propose we merge them in a lockstep. I removed the integration tests labels and added other labels indicating what was changed. This is usually handled by the bot, but it doesn't work with PRs coming from a fork.

jasondavies commented:
@fvranicTT I might take a bit more time to investigate the softmax issue, I believe it's an edge case due to handling of `±inf` inputs, resulting in unexpected results for `1/-inf`.  Should I just mark the PR as draft until resolved?

jasondavies commented:
Ready for review.  Latest updates:

1. Reciprocal now handles values whose absolute bit representation is larger than the magic constant (this is a small range of large floating point values whose reciprocals are very small anyway) and `±inf`, returning `0` for these cases.
2. Similarly, for sqrt/rsqrt, I've added proper handling for both `0` and `+inf`.

These changes add a small number (4-5) of operations to the original routines.  We could of course add various toggles to enable/disable these checks, but I don't think it's worth doing at the moment.

For reciprocal, at some point I could potentially look at adding support for these large floats at the edge of the supported range if necessary, but I think that's probably out of scope for this PR.

And of course, there may be some further optimisations that can be applied, but this is already (a) close to maximal precision and (b) quite optimised already.

For sqrt/rsqrt, negative inputs are not supported right now.  Adding even more checks is possible if desired (e.g. with `nan` for negative inputs) but probably best to handle in a separate PR.  Due to the way that some format conversions mangle `nan` (converting to `inf`) it's not totally straightforward.

jasondavies commented:
@fvranicTT Sorry, noticed an issue with the recently added `_calculate_sqrt_body_` calls in `ckernel_sfpu_trigonometry.h` and just pushed a fix.  I don't know if you need to re-approve to get it into the merge queue or not.

fvranicTT commented:
> @fvranicTT Sorry, noticed an issue with the recently added `_calculate_sqrt_body_` calls in `ckernel_sfpu_trigonometry.h` and just pushed a fix. I don't know if you need to re-approve to get it into the merge queue or not.

@jasondavies I am rerunning the tests now. But we are experiencing some issues with wormhole runners at the moment, so we can’t merge this one (or any other PR) yet. We are working on it.

jasondavies commented:
@fvranicTT Ah, thanks.  I only noticed this particular issue when integration testing with tt-metal locally.  I guess asinh/acosh aren't yet covered by tt-llk tests.

jasondavies commented:
Updates:

- Fixed a register pressure issue in recip (`ttnn/unit_tests/operations/eltwise/test_binaryng_fp32.py::test_pow_fp32`).
- Fixed a self-referential include problem (`ttnn/unit_tests/operations/eltwise/test_binary_bcast.py::test_binary_scalar_ops`).
- Fixed broken BH asinh (unrelated to my changes).

fvranicTT commented:
@jasondavies is this ready to be merged? Do you have a metal counterpart ready?

jasondavies commented:
@fvranicTT Yes, ready to merge. Metal PR is https://github.com/tenstorrent/tt-metal/pull/21623

jasondavies commented:
I took the opportunity to use the new `SFPARECIP` instruction for the BH implementation of reciprocal.

One issue I've noticed is that reciprocal should probably go back to using the `max_iter` template parameter:

- bfloat16 - this can use 0 iterations (<2 ulps) or 1 iteration (0 ulps)
- float16/tf32 - 1 iteration (0 ulps)
- float32 - 2 iterations (<1 ulps)

It's unclear what exactly the use case is for approx_mode, but presumably something like this works:

- approx_mode = true, always use 0 iterations
- approx_mode = false, fp32_dest_acc_en = false: 1 iteration
- approx_mode = false, fp32_dest_acc_en = true: 2 iterations

jasondavies commented:
Reciprocal is now completely rewritten for WH, which shaves off 2 cycles _and_ improves ±0, ±inf handling:

- float32 (max_iter=2): 20 cycles, correct for all normal numbers, ±0, ±inf
- bfloat16/float16 (max_iter=1): 16 cycles, edge case (same as BH) ±8.50705917302346158658e+37 → ±0 instead of ±1.1754943508222875e-38, otherwise correct for all normal numbers, ±0, ±inf

jasondavies commented:
@fvranicTT Ready for another CI run here, thanks.

nmauriceTT commented:
@jasondavies 

The SDXL failures were also happening on [main](https://github.com/tenstorrent/tt-metal/actions/runs/17308917838/job/49177456368) and don't seem related to the PR. And the Models Perf tests failed due to a missing file (also seems unrelated). 

> I do have further optimisation work in the pipeline, but I think the fundamental issue of precision has been resolved by this PR, so assuming everybody is happy with CI on the tt-metal side, it should be good to go.

Yes. That would be great.
